## <center> Знакомимся с данными

Пример задания: 
Ваша задача - разработать программу на языке Python, анализирующую текстовые данные.
Выберите текст, который вам интересен - это может быть:
литературное произведение (например, рассказ, стихотворение, отрывок из романа);
тексты песен;
статьи или посты из открытых источников.
Программа должна считать данные из текстового файла (например, text.txt), выполнить их базовую обработку и вывести результаты анализа.
Пример того, что может делать программа:
посчитать общее количество слов и предложений;
определить 10 самых частотных слов (исключая предлоги, союзы и местоимения);
посчитать, сколько раз встречается выбранное пользователем слово;
вычислить среднюю длину слова и предложения;
посчитать, сколько слов начинается с каждой буквы алфавита;
вывести "облако слов" (в виде текстового списка, без графиков).
Результаты должны быть выведены на экран и записаны в файл results.txt.

Текст произведения мы взяли в библиотеке [lib.ru](http://az.lib.ru/t/tolstoj_lew_nikolaewich/text_0073.shtml) и провели первоначальную обработку. Поскольку наша цель — обработка слов из этого произведения, мы разбили текст на слова и вывели каждое слово в отдельной строке. Кроме того, в местах, где начинаются главы, мы вывели строку `"[new chapter]"`.

> Исходный текстовый файл хранится в общем доступе и находится [здесь](https://raw.githubusercontent.com/SkillfactoryDS/Datasets/master/war_peace_processed.txt).

Для начала скачаем текст книги по ссылке.

In [63]:
# Импортируем библиотеку для выполнения HTTP-запросов в интернет
import requests

# Читаем текстовый файл по url-ссылке
data = requests.get("https://raw.githubusercontent.com/SkillfactoryDS/Datasets/master/war_peace_processed.txt").text

# Предобрабатываем текстовый файл
data = data.split('\n')
data.remove('')
data = data + ['[new chapter]']

# Выводим первые 100 слов из книги
print(data[:100])

['1', 'в', 'два', 'раза', 'короче', 'и', 'в', 'пять', 'раз', 'интереснее', '2', 'почти', 'нет', 'философических', 'отступлений', '3', 'в', 'сто', 'раз', 'легче', 'читать', 'весь', 'французский', 'текст', 'заменен', 'русским', 'в', 'переводе', 'самого', 'толстого', '4', 'гораздо', 'больше', 'мира', 'и', 'меньше', 'войны', '5', 'хеппи-энд', 'эти', 'слова', 'я', 'поместил', 'семь', 'лет', 'назад', 'на', 'обложку', 'предыдущего', 'издания', 'указав', 'в', 'аннотации', 'первая', 'полная', 'редакция', 'великого', 'романа', 'созданная', 'к', 'концу', '1866', 'года', 'до', 'того', 'как', 'толстой', 'переделал', 'его', 'в', '1867--1869', 'годах', '--', 'и', 'что', 'я', 'использовал', 'такие-то', 'публикации', 'думая', 'что', 'все', 'всё', 'знают', 'я', 'не', 'объяснил', 'откуда', 'взялась', 'эта', 'первая', 'редакция', 'я', 'оказался', 'неправ', 'и', 'в', 'результате', 'оголтелые', 'и']


## <center> Работаем с данными

Для начала найдем общее количество слов и количество уникальных слов в тексте

In [64]:
# Превращаем список в множество, удаляя дублирующиеся слова
word_set = set(data)
print(len(word_set))
# Удаляем из множества слово, символизирующее раздел между главами
word_set.discard('[new chapter]')
# word_set.discard()
# Выводим результаты
print('Общее количество слов: {}'.format(len(data)))
print('Общее количество уникальных слов: {}'.format(len(word_set)))

38211
Общее количество слов: 300080
Общее количество уникальных слов: 38210


Давайте напишем программу, которая посчитает частоту каждого слова. Для этого создадим словарь, ключами которого будут являться слова, а значения - количество вхождений этого слова в текст произведения. Заодно подсчитаем количество глав

In [65]:
# Инициализируем пустой словарь
word_counts = {}
# Инициализируем количество глав
count_chapter = 0
# Создаем цикл по всем словам из списка слов
for word in data:
    # Проверяем, что текущее слово - обозначение новой главы
    if word == '[new chapter]':
        # Если условие выполняется, то увеличиваем количество глав на 1
        count_chapter += 1
        # Переходим на новую итерацию цикла
        continue
    # Проверяем, что текущего слова еще нет в словаре слов
    if word not in word_counts:
        # Если условие выполняется, инициализируем новый ключ 1
        word_counts[word] = 1
    else:
        # В противном случае, увеличиваем количество слов на 1
        word_counts[word] += 1

# Выводим количество глав
print('Количество глав: {}'.format(count_chapter))

# Создаем цикл по ключам и их порядковым номерам полученного словаря
for i, key in enumerate(word_counts):
    # Выводим только первые 10 слов
    if i == 10:
        break
    print(key, word_counts[key])

Количество глав: 171
1 10
в 6997
два 215
раза 35
короче 3
и 14592
пять 61
раз 296
интереснее 4
2 12


In [66]:
# Создаем цикл по ключам и их порядковым номерам полученного словаря
for i, key in enumerate(word_counts):
    # Выводим только первые 10 слов
    # if i == 10:
    #     break
    print(key, word_counts[key])

1 10
в 6997
два 215
раза 35
короче 3
и 14592
пять 61
раз 296
интереснее 4
2 12
почти 86
нет 460
философических 1
отступлений 1
3 8
сто 19
легче 12
читать 35
весь 127
французский 36
текст 4
заменен 1
русским 9
переводе 1
самого 71
толстого 15
4 6
гораздо 32
больше 233
мира 38
меньше 38
войны 62
5 5
хеппи-энд 1
эти 263
слова 177
я 2228
поместил 1
семь 15
лет 77
назад 158
на 4164
обложку 1
предыдущего 1
издания 3
указав 3
аннотации 1
первая 23
полная 7
редакция 2
великого 33
романа 6
созданная 1
к 2235
концу 4
1866 3
года 76
до 424
того 451
как 2941
толстой 12
переделал 3
его 2692
1867--1869 1
годах 3
-- 9680
что 5097
использовал 2
такие-то 4
публикации 1
думая 17
все 1891
всё 14
знают 29
не 5922
объяснил 15
откуда 14
взялась 7
эта 132
оказался 1
неправ 5
результате 2
оголтелые 1
невежественные 1
критики 2
выдающие 1
себя 542
за 1232
знатоков 1
русской 24
литературы 1
публично 2
стали 81
обвинять 1
меня 455
фальсификации 1
это 1851
захаров 2
сам 311
сляпал 1
надругательстве 1
над 198
толс

Разделим все слова на главы. Для этого создадим список, в котором будем хранить списки - слова из определенной главы.

In [67]:
# Инициализируем общий список, в котором будем хранить списки слов в каждой главе
chapter_data = []
# Инициализируем список слов, в котором будет хранить слова одной главы
chapter_words = []

# Создаем цикл по всем словам из списка
for word in data:
    # Проверяем, что текущее слово - обозначение новой главы
    if word == '[new chapter]':
        # Если условие выполняется, добавляем список со словами из главы в общий список
        chapter_data.append(chapter_words)
        # Обновляем (перезаписываем) список со словами из текущей главы
        chapter_words = []
    else:
        # В противном случае, добавляем текущее слово в список со словами из главы
        chapter_words.append(word)

# Проверяем, что у нас получилось столько же списков, сколько глав в произведении
print('Вложенный список содержит {} внутренних списка'.format(len(chapter_data)))
# Выведем первые 100 слов 0-ой главы
print(chapter_data[0][:100])

Вложенный список содержит 171 внутренних списка
['1', 'в', 'два', 'раза', 'короче', 'и', 'в', 'пять', 'раз', 'интереснее', '2', 'почти', 'нет', 'философических', 'отступлений', '3', 'в', 'сто', 'раз', 'легче', 'читать', 'весь', 'французский', 'текст', 'заменен', 'русским', 'в', 'переводе', 'самого', 'толстого', '4', 'гораздо', 'больше', 'мира', 'и', 'меньше', 'войны', '5', 'хеппи-энд', 'эти', 'слова', 'я', 'поместил', 'семь', 'лет', 'назад', 'на', 'обложку', 'предыдущего', 'издания', 'указав', 'в', 'аннотации', 'первая', 'полная', 'редакция', 'великого', 'романа', 'созданная', 'к', 'концу', '1866', 'года', 'до', 'того', 'как', 'толстой', 'переделал', 'его', 'в', '1867--1869', 'годах', '--', 'и', 'что', 'я', 'использовал', 'такие-то', 'публикации', 'думая', 'что', 'все', 'всё', 'знают', 'я', 'не', 'объяснил', 'откуда', 'взялась', 'эта', 'первая', 'редакция', 'я', 'оказался', 'неправ', 'и', 'в', 'результате', 'оголтелые', 'и']


In [68]:
chapter_data[15][100]

'в'

Подсчитаем, сколько раз каждое слово встречается в каждой из глав

In [69]:
# Инициализируем список, в котором будем хранить словари
chapter_words_count = []

# Создаем цикл по элементам внешнего списка со словами
for chapter_words in chapter_data:
    # Инициализируем пустой словарь, куда будем добавлять результаты
    temp = {}
    # Создаем цикл по элементам внутреннего списка
    for word in chapter_words:
        # Проверяем, что текущего слова еще нет в словаре
        if word not in temp:
            # Если условие выполняется, добавляем ключ в словарь
            temp[word] = 1
        else:
            # В противном случае, увеличиваем количество влождений слова в главу
            temp[word] += 1
    # Добавляем получившийся словарь в список
    chapter_words_count.append(temp)

# Выводим результат
print(chapter_words_count)

[{'1': 1, 'в': 37, 'два': 3, 'раза': 1, 'короче': 1, 'и': 34, 'пять': 1, 'раз': 2, 'интереснее': 1, '2': 1, 'почти': 1, 'нет': 1, 'философических': 1, 'отступлений': 1, '3': 1, 'сто': 1, 'легче': 1, 'читать': 1, 'весь': 2, 'французский': 2, 'текст': 3, 'заменен': 1, 'русским': 1, 'переводе': 1, 'самого': 2, 'толстого': 5, '4': 1, 'гораздо': 1, 'больше': 1, 'мира': 3, 'меньше': 1, 'войны': 3, '5': 1, 'хеппи-энд': 1, 'эти': 2, 'слова': 2, 'я': 8, 'поместил': 1, 'семь': 1, 'лет': 2, 'назад': 1, 'на': 7, 'обложку': 1, 'предыдущего': 1, 'издания': 3, 'указав': 1, 'аннотации': 1, 'первая': 2, 'полная': 1, 'редакция': 2, 'великого': 1, 'романа': 5, 'созданная': 1, 'к': 4, 'концу': 2, '1866': 3, 'года': 5, 'до': 1, 'того': 1, 'как': 2, 'толстой': 6, 'переделал': 1, 'его': 5, '1867--1869': 1, 'годах': 2, '--': 10, 'что': 9, 'использовал': 2, 'такие-то': 1, 'публикации': 1, 'думая': 1, 'все': 2, 'всё': 7, 'знают': 1, 'не': 11, 'объяснил': 1, 'откуда': 1, 'взялась': 1, 'эта': 1, 'оказался': 1, 'н

In [70]:
chapter_words_count[15]['князю']

2

In [71]:
# Создаем цикл по ключам словаря - спискам слов и их порядковым номерам
for chapter_number, chapter_dict in enumerate(chapter_words_count):
    # Выводим только первые 5 глав
    if chapter_number == 5:
        break
    # Выводим номер главы
    print('-' * 40)
    print('Chapter: {}'.format(chapter_number))
    print('-' * 40)
    # Создаем цикл по ключам - словам и их порядковым номерам
    for j, word in enumerate(chapter_dict):
        # Выводим первые 10 слов из главы
        if j == 10:
            break
        print(word, chapter_dict[word])

----------------------------------------
Chapter: 0
----------------------------------------
1 1
в 37
два 3
раза 1
короче 1
и 34
пять 1
раз 2
интереснее 1
2 1
----------------------------------------
Chapter: 1
----------------------------------------
автора 1
я 20
пишу 2
до 1
сих 1
пор 1
только 5
о 2
князьях 1
графах 1
----------------------------------------
Chapter: 2
----------------------------------------
первая 1
----------------------------------------
Chapter: 3
----------------------------------------
-- 81
ну 5
что 44
князь 21
генуя 1
и 94
лукка 1
стали 1
не 57
больше 2
----------------------------------------
Chapter: 4
----------------------------------------
гостиная 1
анны 2
павловны 2
начала 1
понемногу 1
наполняться 1
приехала 3
высшая 1
знать 1
петербурга 2


Давайте резюмировать, что мы с вами уже получили:

* `word_set` - множество из всех слов, которые есть в книге

* `count_chapter` - количество глав в книге (171)

* `word_counts` - словарь, ключами которого являются слова, а значениями - количество вхождений этих слов в книгу

* `chapter_data` - список из 171 списка, где элементы вложенных списков - все слова из главы. Каждый список соответствует своей главе

* `chapter_words_count` - список из 171 словаря, где ключи - слова, а значения - количество слов в главе. Каждый словарь соответствует своей главе

Учтите, что эти данные могут пригодиться вам при выполнении дальнейших заданий.

> А теперь к заданиями!

## <center> Задания для самостоятельного решения

### Задание 1.

Давайте введем понятие частоты употребления отдельного слова в документе (`term frequency`, или `tf`). В нашем случае речь идёт не о документах, а о главах книги (выше мы писали, что в текстовом документе главы разделяются строкой '[new chapter]').

Формула для вычисления `term frequency` для слова `word`:
$$ tf_{word, chapter} = \frac {n_{word, chapter}} {n_{chapter}}$$

где
* ${n_{word, chapter}}$ - сколько раз слово `word` встречается в главе `chapter`,
* $n_{chapter}$ - количество слов в главе `chapter`.


Например, слово `"гостья"` употребляется в 15-ой главе 10 раз (${n_{word, chapter}}$).(кстати, главы у нас нумеруются с 0). Общее количество слов в тексте 15-ой главы - 1359 ($n_{chapter}$). Тогда:

$$ tf_{гостья, 15} = \frac{10}{1359} \approx 0.007358$$

**Задание:**

Напишите программу, которая позволит получать частоту употребления любого заданного слова `target_word` в заданной главе `target_chapter`.

**Дополнительное требование:**

*Пострайтесь сделать программу максимально обобщенной. То есть желательно рассчитать характеристику `tf` для всех слов из каждой главы, чтобы впоследствии не было необходимости производить вычисления снова.*

**Подсказка:**

*Для этого вы можете для каждой главы создать словарь, ключами которого являются слова, а значения - частота употребления этого слова в этой главе*

**Протестируйте работу программы на нескольких словах и главах.**

In [72]:
target_word = 'гостья'
target_chapter = 15

tf=chapter_words_count[target_chapter][target_word]/len(chapter_data[15])
round(tf,6)

0.007358

In [73]:
target_word = 'в'
target_chapter = 168

tf=chapter_words_count[target_chapter][target_word]/len(chapter_data[15])
round(tf,6)

0.025018

In [74]:
# ${n_{word, chapter}}$ - сколько раз слово `word` встречается в главе `chapter`,

chapter_words_count[15]['гостья']

10

In [75]:
# * $n_{chapter}$ - количество слов в главе `chapter`.
len(chapter_data[15])

1359

In [76]:
# создаем пустой список в который будем вносить словари, в которых ключ-это слово главы, а значения - частота употребления  это слова в данной главе
term_frequency_for_all_dict=[]
for chapter_number,words_in_chapter in enumerate(chapter_words_count):
    # print (chapter_number,words_in_chapter)
    temp={}
    for i,j in enumerate(words_in_chapter):
        # print(i,j)
        target_word_quantity=words_in_chapter[j]
        # print(target_word_quantity)
        all_word_chapter_quantity=len(chapter_data[chapter_number])
        # print(all_word_chapter_quantity)
        term_frequency=target_word_quantity/all_word_chapter_quantity
        # print(term_frequency)
        temp[j]=term_frequency
    term_frequency_for_all_dict.append(temp)
    # print(term_frequency_for_all_dict)
print(term_frequency_for_all_dict[target_chapter][target_word])


0.03168685927306617


In [77]:
target_word = 'гостья'
target_chapter = 15

# Ваш код здесь

### Задание 2.

Пришло время познакомиться с понятием `document frequency`.

`Document frequency` (для удобства сократим до `df`) — это доля документов, в которых встречается искомое слово.

Вычисляется по формуле:

$$ df_{word} = \frac{N_{word}}{N} $$,

где
* $N_{word}$ - число документов (глав) содержащих слово `word`,
* $N$ - общее число документов (глав).

Объясним на примере: наш текст состоит из 171 главы ($N$), а слово `"человек"` встречается в 115 главах. Тогда:

$$ df_{человек} = \frac{115}{171} \approx 0.6725$$

**Задание:**

Напишите программу, которая позволит вычислять document frequency для заданного слова `target_word` и выведить результат на экран.

**Дополнительное требование:**

*Пострайтесь сделать программу максимально обобщенной. То есть желательно рассчитать характеристику `df` для всех уникальных слов из книги, чтобы впоследствии не было необходимости производить вычисления снова.*

**Подсказка:**
*Для этого вы можете создать словарь, ключами которого являются слова из книги, а значения - доля документов, содержащих эти слова*

**Протестируйте работу программы на нескольких словах**

In [78]:
target_word = 'лес'

# Ваш код здесь
word_quantity_dict={} #создаем пустой словарь
for words in word_set: #создаем цил по словам ранне полученного множества word_set
    quantity_word=0 # создаем счетчик слов во множестве word_set
    for chapter, i in enumerate(chapter_words_count):# создаем цикл по главам и словам ранее полученного списка chapter_words_count
        if words in chapter_words_count[chapter]:  # создаем условие : если слова есть в ключах словаря в списке  chapter_words_count
            quantity_word+=1 # увеличиваем счетчик на единицу
        else: # если условие не выполняется - пропускаем итерацию
            pass
    word_quantity_dict[words]=quantity_word # добавляем в пустой словарь у которого ключ= слово, значение равное количеству слов из счетчика
document_frequency_all_dict={} # иницируем пустой словарь, ключи которого - слова из книги, а значения - доля докуметов, содержащих эти слова
for words in word_quantity_dict:# создаем цикл для добавления в словарь значений
    document_frequency_all_dict[words]=word_quantity_dict[words]/len(chapter_words_count) # находим значения (доля документов содержащих эти слова) словаря

print(document_frequency_all_dict[target_word]) #печатаем на экран искомое значение  целевого слова target_word

0.04678362573099415


### Задание 3

Пришло время дать разъяснения: для чего мы делали вычисления выше и что нас ждет впереди?

> Если какое-то слово часто употребляется в документе, то, вероятно, этот документ что-то рассказывает о предмете/действии, описываемом этим словом. Скажем, если вы читаете книгу, в которой много раз употребляется слово `"заяц"`, то, вероятно, эта книга про зайцев.

> Однако, если вы возьмёте слово `"и"`, то оно будет встречаться почти в каждой книге много раз.

Таким образом, если мы хотим найти наиболее значимые слова в книге, мы, с одной стороны, хотим найти наиболее частые слова, а с другой — убрать те, которые не несут важной информации, так как встречаются везде.

Такая задача хорошо решается с помощью `tf-idf` — статистической метрики для оценки важности слова в тексте. Другими словами, `tf-idf` — это «контрастность» слова в документе (насколько оно выделяется среди других слов).

Формула для вычисления следующая:

`tf-idf = term frequency * inverse document frequency`

* `tf` — это частотность термина, которая измеряет, насколько часто термин встречается в документе.

* `idf` — это обратная документная частотность термина. Она измеряет непосредственно важность термина во всём множестве документов.

Чтобы получить `idf`, необходимо поделить 1 на полученную в Задании 2 документную частоту (`df`):

$$idf = \frac{1}{df}$$

Мы будем использовать не сырые значения `idf`, а их логарифмы, то есть $tf * log(idf)$. Сейчас мы не будем заострять внимания на том, почему следует использовать именно логарифм — это долгий разговор. Вернемся к нему, когда будем изучать методы машинного обучения для обработки текстов. Подробнее о `tf-idf` вы можете почитать [здесь](https://www.freecodecamp.org/news/how-to-process-textual-data-using-tf-idf-in-python-cd2bbc0a94a3/).

В качестве примера измерим `tf-idf` слова `"анна"` в главе 4. Слово `"анна"` встречается в указанной главе 7 раз, при этом в 4 главе 1060 слов, всего же слово `"анна"` упоминается в 32 главах из 171.

Таким образом, `tf-idf` данного слова в данной главе будет равно:

$$tf\_idf_{анна, 4} = tf * log(\frac{1}{df}) = \frac{7}{1060} * log(\frac {171}{32}) \approx 0.011067$$

**Примечание:** здесь используется натуральный логарифм по основанию $e$, однако в общем случае основание логарифма не имеет значения, так как характеристика `tf-idf` используется для сравнения контрастности слов между собой

**Задание**:

Напишите программу, которая позволяет вычислять значение `tf-idf` для заданного слова `target_word` в заданной главе `target_chapter`.

**Дополнительное требование:**

*Пострайтесь сделать программу максимально оптимальной. То есть желательно рассчитать характеристику `tf-idf` для всех слов из каждой главы книги, чтобы впоследствии не было необходимости производить вычисления снова.*

**Протестируйте работу программы на нескольких словах и главах.**

**Примечание:**

Натуральный логарифм можно вычислить с помощью функции [log](https://pythonim.ru/chisla/funktsiya-log-v-python) из встроенного в Python модуля math:


In [79]:
# Импортируем функцию log из модуля math:
from math import log
print(log(18))

2.8903717578961645


**Примечание.**

**Модуль (библиотека) в Python** — это любой программный файл, который содержит в себе код, включая функции. В нашем случае math — это встроенный модуль, содержащий функционал для математических вычислений. Подробнее о math вы можете почитать [здесь](https://pythonworld.ru/moduli/modul-math.html).

In [80]:
target_word = 'анна'
target_chapter = 4

# Ваш код здесь
from math import log

from math import log
tf_idf_for_all_dict=[]# создаем пустой список
for chapter_number, word in enumerate(chapter_data):#создаем цикл по словам ранее полученного списка chapter_data
    temp_tf_idf_dict={} # создаем пустой промежуточный словарь для записи результатов
    for i, j in enumerate(word): # создаем внутренний цикл по словарям в списке
        tf_idf=(term_frequency_for_all_dict[chapter_number][j])*(log(1/document_frequency_all_dict[j])) # находим значения для переменной tf_idf
        temp_tf_idf_dict[j]=tf_idf # добавляем найденные значения в словарь
    tf_idf_for_all_dict.append(temp_tf_idf_dict) # добавляем полученные словари в список tf_idf_dic_all---tf_idf_for_all_dict
print(tf_idf_for_all_dict[target_chapter][target_word]) #Выводим результат для целевого слова target_word и целевой главы target_chapter

0.011067446769736353


### Задание 4.

Теперь, когда мы умеем вычислять `tf-idf` для каждого слова в главе, мы можем найти те слова, которые являются самыми «контрастными» для данной главы, то есть они могут являться в своём роде заголовком для главы.

Например, для главы 3 наиболее значимыми словами будут:

`"анна", "павловна", "функе"`

**Задание:**

Напишите программу, которая позволяет вывести три слова, имеющие самое высокое значение `tf-idf` в заданной главе `target_chapter` в порядке убывания `tf-idf`.

In [81]:
target_chapter = 3

# Ваш код здесь
list_for_target_chapter=tf_idf_for_all_dict[target_chapter]# определяем список для анализируемой главы
first_word = sorted(list_for_target_chapter, key=list_for_target_chapter.get)[-1]# сортируем словарь и находим слово с максимальной частотой
second_word = sorted(list_for_target_chapter, key=list_for_target_chapter.get)[-2]# сортируем словарь и находим второе по применяемости слово
third_word = sorted(list_for_target_chapter, key=list_for_target_chapter.get)[-3]# сортируем словарь и находим третье по применяемости слово
print(first_word) # выводим на печать первое слово
print(second_word)# выводим на печать второе слово
print(third_word)# выводим на печать третье слово

павловна
анна
функе


!

Пример задания: 

Ваша задача - разработать программу на языке Python, анализирующую текстовые данные.

Выберите текст, который вам интересен - это может быть: литературное произведение (например, рассказ, стихотворение, отрывок из романа); тексты песен; статьи или посты из открытых источников.

Программа должна считать данные из текстового файла (например, text.txt), выполнить их базовую обработку и вывести результаты анализа.
Пример того, что может делать программа:
посчитать общее количество слов и предложений;
определить 10 самых частотных слов (исключая предлоги, союзы и местоимения);
посчитать, сколько раз встречается выбранное пользователем слово;
вычислить среднюю длину слова и предложения;
посчитать, сколько слов начинается с каждой буквы алфавита;
вывести "облако слов" (в виде текстового списка, без графиков).
Результаты должны быть выведены на экран и записаны в файл results.txt.

Условие задачи. Подсчитать количество вхождений каждого символа в заданном тексте. В результате работы программы должен быть сформирован словарь, ключи которого — символы текста, а значения — количество вхождений символа в тексте.

In [82]:
# Импортируем библиотеку для выполнения HTTP-запросов в интернет
import requests

# Читаем текстовый файл по url-ссылке
data = requests.get("https://raw.githubusercontent.com/SkillfactoryDS/Datasets/master/war_peace_processed.txt").text

# Предобрабатываем текстовый файл
data = data.split('\n')
data.remove('')
data = data + ['[new chapter]']

# Выводим первые 100 слов из книги
print(data[:100])

['1', 'в', 'два', 'раза', 'короче', 'и', 'в', 'пять', 'раз', 'интереснее', '2', 'почти', 'нет', 'философических', 'отступлений', '3', 'в', 'сто', 'раз', 'легче', 'читать', 'весь', 'французский', 'текст', 'заменен', 'русским', 'в', 'переводе', 'самого', 'толстого', '4', 'гораздо', 'больше', 'мира', 'и', 'меньше', 'войны', '5', 'хеппи-энд', 'эти', 'слова', 'я', 'поместил', 'семь', 'лет', 'назад', 'на', 'обложку', 'предыдущего', 'издания', 'указав', 'в', 'аннотации', 'первая', 'полная', 'редакция', 'великого', 'романа', 'созданная', 'к', 'концу', '1866', 'года', 'до', 'того', 'как', 'толстой', 'переделал', 'его', 'в', '1867--1869', 'годах', '--', 'и', 'что', 'я', 'использовал', 'такие-то', 'публикации', 'думая', 'что', 'все', 'всё', 'знают', 'я', 'не', 'объяснил', 'откуда', 'взялась', 'эта', 'первая', 'редакция', 'я', 'оказался', 'неправ', 'и', 'в', 'результате', 'оголтелые', 'и']


In [160]:
# Программа для чтения всего файла с использованием функции read()
file = open("C:\скил\DST-156\IDE_NEW\PYTHON-4. Циклы/Text Document.txt", "r")
text = file.read()
file.close()


print(text)

CHAPTER I
Down the Rabbit-Hole
Alice was beginning to get very tired of sitting by her sister on the bank, and of having nothing to do: once or twice she had peeped into the book her sister was reading, but it had no pictures or conversations in it, 'and what is the use of a book,' thought Alice 'without pictures or conversation?'
So she was considering in her own mind (as well as she could, for the hot day made her feel very sleepy and stupid), whether the pleasure of making a daisy-chain would be worth the trouble of getting up and picking the daisies, when suddenly a White Rabbit with pink eyes ran close by her.
There was nothing so very remarkable in that; nor did Alice think it so very much out of the way to hear the Rabbit say to itself, 'Oh dear! Oh dear! I shall be late!' (when she thought it over afterwards, it occurred to her that she ought to have wondered at this, but at the time it all seemed quite natural); but when the Rabbit actually took a watch out of its waistcoat-po

Подсчитаем колличествоо вхождений каждого символа в заданном тексте. В качестве результата сформируем словарь, ключи которого — символы текста, а значения — количество вхождений символа в тексте.
Условимся, что символы в верхнем и нижнем регистре будем считать одинаковыми, поэтому приведём текст в нижний регистр и удалим все пробелы и символы переноса строки ('\n')
Чтобы привести текст в нижний регистр, используем метод lower(). 

Для того чтобы заменить пробелы и символы переноса строки ('\n') используем метод replace(): первым аргументом в него передаётся символ, который необходимо заменить, вторым аргументом — то, на что его надо заменить. Пробелы и перенос строки заменяем на пустые строки '' (то есть на «ничего»).

Метод lower() и replace() мы уже изучали, когда говорили о строковых типах данных.

Обратите внимание: так как методы lower() и replace() возвращают преобразованные копии текста, но не изменяют сам текст, нам необходимо переопределить переменную text.

# Приводим текст к нижнему регистру
text = text.lower() 
# Заменяем пробелы на пустые строки
text = text.replace(" ", "")
# Заменяем символы переноса строки на пустые строки 
text = text.replace("\n", "") 
# Выводим результирующий текст
print(text)

In [161]:
text=text.lower() # Приводим текст к нижнему регистру
text=text.replace(' ','') # Заменяем пробелы на пустые строки
text=text.replace('\n','') # Заменяем символы переноса строки на пустые строки 
# Выводим результирующий текст
#print(text)
count_dict={}
for simbol in text:
    if simbol not in count_dict:
        count_dict[simbol]=1
        # print(count_dict)
    else:
        count_dict[simbol]+=1
        # print(count_dict)
sorted(count_dict.items())

[('!', 127),
 ("'", 350),
 ('(', 23),
 (')', 22),
 (',', 531),
 ('-', 60),
 ('.', 149),
 (':', 56),
 (';', 39),
 ('?', 37),
 ('a', 1990),
 ('b', 338),
 ('c', 503),
 ('d', 1126),
 ('e', 2863),
 ('f', 469),
 ('g', 561),
 ('h', 1588),
 ('i', 1712),
 ('j', 19),
 ('k', 235),
 ('l', 1157),
 ('m', 468),
 ('n', 1548),
 ('o', 1934),
 ('p', 332),
 ('q', 34),
 ('r', 1187),
 ('s', 1484),
 ('t', 2232),
 ('u', 744),
 ('v', 172),
 ('w', 667),
 ('x', 20),
 ('y', 500),
 ('z', 17)]

In [164]:
# Программа для чтения всего файла с использованием функции read()
file = open("C:\скил\DST-156\IDE_NEW\PYTHON-4. Циклы/Text Document.txt", "r")
text = file.read()
file.close()
# Приводим текст к нижнему регистру
text = text.lower() 
# Заменяем символы переноса строки на пробелы
text = text.replace("\n", " ") 

text = text.replace("!", "")
text = text.replace("'", "")
text = text.replace("(", "")
text = text.replace(")", "")
text = text.replace(",", "")
text = text.replace(";", "")
text = text.replace("-", "")
text = text.replace(":", "")
text = text.replace("?", "")
text = text.replace(".", "")
text = text.replace("-", "")
text = text.replace("?", "")


# Разделяем текст на слова
word_list = text.split()
# Создаём пустой словарь для подсчёта количества слов
count_dict = {} 
# Создаём цикл по словам в списке word_list
# word — текущее слово из списка word_list
for word in word_list: 
    # Проверяем условие, что слова ещё нет среди ключей словаря.
    if word not in count_dict:
        # Если условие выполняется, заносим слово в словарь со значением 1.
        count_dict[word] = 1
    else: 
        # В противном случае увеличиваем частоту слова
        count_dict[word] += 1
# Выводим результирующий словарь
print(count_dict)
sorted(count_dict.items()) 

{'chapter': 3, 'i': 145, 'down': 40, 'the': 282, 'rabbithole': 3, 'alice': 76, 'was': 107, 'beginning': 1, 'to': 189, 'get': 13, 'very': 38, 'tired': 4, 'of': 102, 'sitting': 1, 'by': 14, 'her': 65, 'sister': 2, 'on': 40, 'bank': 1, 'and': 209, 'having': 4, 'nothing': 10, 'do': 29, 'once': 10, 'or': 19, 'twice': 1, 'she': 172, 'had': 47, 'peeped': 1, 'into': 14, 'book': 3, 'reading': 1, 'but': 45, 'it': 143, 'no': 22, 'pictures': 4, 'conversations': 1, 'in': 86, 'what': 24, 'is': 62, 'use': 7, 'a': 146, 'thought': 22, 'without': 4, 'conversation': 2, 'so': 34, 'considering': 2, 'own': 4, 'mind': 4, 'as': 53, 'well': 10, 'could': 24, 'for': 45, 'hot': 3, 'day': 2, 'made': 5, 'feel': 2, 'sleepy': 2, 'stupid': 2, 'whether': 2, 'pleasure': 1, 'making': 2, 'daisychain': 1, 'would': 21, 'be': 41, 'worth': 2, 'trouble': 2, 'getting': 8, 'up': 25, 'picking': 1, 'daisies': 1, 'when': 22, 'suddenly': 6, 'white': 6, 'rabbit': 10, 'with': 39, 'pink': 1, 'eyes': 6, 'ran': 3, 'close': 3, 'there': 20

[('a', 146),
 ('able', 1),
 ('about', 23),
 ('absurd', 1),
 ('acceptance', 1),
 ('accustomed', 1),
 ('across', 2),
 ('actually', 1),
 ('ada', 1),
 ('added', 1),
 ('addressing', 1),
 ('adjourn', 1),
 ('adoption', 1),
 ('advice', 1),
 ('advisable', 2),
 ('advise', 1),
 ('afraid', 4),
 ('after', 11),
 ('afterwards', 1),
 ('again', 20),
 ('against', 2),
 ('age', 2),
 ('ago', 1),
 ('ah', 2),
 ('ahem', 1),
 ('air', 3),
 ('alas', 3),
 ('alice', 76),
 ('alive', 1),
 ('all', 44),
 ('allow', 1),
 ('almost', 3),
 ('alone', 2),
 ('along', 3),
 ('aloud', 2),
 ('altogether', 2),
 ('always', 3),
 ('am', 20),
 ('among', 4),
 ('an', 8),
 ('and', 209),
 ('andoh', 2),
 ('angrily', 1),
 ('angry', 1),
 ('animal', 1),
 ('animals', 2),
 ('another', 4),
 ('answer', 3),
 ('antipathies', 1),
 ('anxiously', 3),
 ('any', 10),
 ('anything', 5),
 ('apiece', 1),
 ('archbishop', 2),
 ('are', 11),
 ('argument', 1),
 ('as', 53),
 ('ashamed', 1),
 ('ask', 4),
 ('asked', 1),
 ('asking', 2),
 ('assembled', 1),
 ('at', 31)